File: apps/api/app/models/table.py
"""Table-related Pydantic models."""
from enum import Enum

from pydantic import BaseModel


class ColumnFormat(str, Enum):
    """Column format options."""
    TEXT = "text"
    DATE = "date"
    TIMERANGE = "timerange"


class CreateTableRequest(BaseModel):
    """Request model for creating a new table."""

    title: str | None = None
    description: str | None = None
    cols: int = 4
    rows: int = 10


class CreateTableResponse(BaseModel):
    """Response model for table creation."""

    slug: str
    admin_token: str
    edit_token: str


class TableColumn(BaseModel):
    """Table column model."""

    idx: int
    header: str | None
    width: int | None
    format: ColumnFormat = ColumnFormat.TEXT


class CellData(BaseModel):
    """Cell data model."""

    row: int
    col: int
    value: str | None


class TableResponse(BaseModel):
    """Response model for table data."""

    id: str
    slug: str
    title: str | None
    description: str | None
    cols: int
    rows: int
    fixed_rows: bool = False
    columns: list[TableColumn]
    cells: list[CellData] = []


class CellUpdateRequest(BaseModel):
    """Request model for updating cells."""

    row: int
    col: int
    value: str | None = None


class CellBatchUpdateRequest(BaseModel):
    """Request model for batch cell updates."""

    cells: list[CellUpdateRequest]


class ColumnConfigUpdate(BaseModel):
    """Request model for updating column configuration."""

    idx: int
    header: str | None = None
    width: int | None = None
    format: ColumnFormat | None = None


class TableConfigRequest(BaseModel):
    """Request model for updating table configuration."""

    title: str | None = None
    description: str | None = None
    rows: int | None = None
    fixed_rows: bool | None = None
    columns: list[ColumnConfigUpdate] | None = None


class TableConfigResponse(BaseModel):
    """Response model for table configuration."""

    success: bool
    message: str
    limits: dict[str, int] = {}


class AddRowRequest(BaseModel):
    """Request model for adding rows to a table."""

    count: int = 1


class RemoveRowRequest(BaseModel):
    """Request model for removing rows from a table."""

    count: int = 1


class AddColumnRequest(BaseModel):
    """Request model for adding columns to a table."""

    count: int = 1
    header: str | None = None


class RemoveColumnRequest(BaseModel):
    """Request model for removing columns from a table."""

    count: int = 1


class RowColumnResponse(BaseModel):
    """Response model for row/column operations."""

    success: bool
    message: str
    new_rows: int | None = None
    new_cols: int | None = None

File: apps/web/src/types/table.ts
/**
 * Table-related TypeScript types.
 */

export type ColumnFormat = 'text' | 'date'

export interface TableColumn {
  idx: number
  header: string | null
  width: number | null
  format: ColumnFormat
}

export interface CellData {
  row: number
  col: number
  value: string | null
}

export interface TableData {
  id: string
  slug: string
  title: string | null
  description: string | null
  cols: number
  rows: number
  fixed_rows: boolean
  columns: TableColumn[]
  cells: CellData[]
}

export interface CreateTableRequest {
  title?: string
  description?: string
  cols?: number
  rows?: number
}

export interface CreateTableResponse {
  slug: string
  admin_token: string
  edit_token: string
}

export interface CellUpdateRequest {
  row: number
  col: number
  value: string | null
}

export interface CellBatchUpdateRequest {
  cells: CellUpdateRequest[]
}

export type UserRole = 'admin' | 'editor'

export interface LoadingState {
  isLoading: boolean
  error: string | null
}

export interface ColumnConfigUpdate {
  idx: number
  header?: string | null
  width?: number | null
  format?: ColumnFormat | null
}

export interface TableConfigRequest {
  title?: string | null
  description?: string | null
  rows?: number | null
  fixed_rows?: boolean | null
  columns?: ColumnConfigUpdate[] | null
}

export interface TableConfigResponse {
  success: boolean
  message: string
  limits: {
    max_rows: number
    max_cols: number
  }
}

export interface AddRowRequest {
  count?: number
}

export interface RemoveRowRequest {
  count?: number
}

export interface AddColumnRequest {
  count?: number
  header?: string
}

export interface RemoveColumnRequest {
  count?: number
}

export interface RowColumnResponse {
  success: boolean
  message: string
  new_rows?: number | null
  new_cols?: number | null
}
